File: offer_1_0036/cpp/offer_1_0036.cpp
//
// Created by wangzy on 2022/4/6.
//

#include<iostream>
#include<queue>
#include<vector>
#include <algorithm>
#include <unordered_map>
#include <unordered_set>
#include <list>
#include <deque>
#include <stack>
#include <sstream>
#include <typeinfo>
#include <climits>
#include <random>
#include <ctime>

using namespace std;

class Node {
public:
    int val;
    Node* left;
    Node* right;

    Node() {}

    Node(int _val) {
        val = _val;
        left = NULL;
        right = NULL;
    }

    Node(int _val, Node* _left, Node* _right) {
        val = _val;
        left = _left;
        right = _right;
    }
};

class Solution {
public:
    Node* treeToDoublyList(Node* root) {
        Node* head = nullptr;
        Node* pre = nullptr;
        if(!root)
            return nullptr;
        //中序遍历，保留前一个节点，修改当前节点的左子树指向(左子树已经遍历完可修改)和前一个节点的右子树指向
        inOrderHelper(root,head,pre);
        //最后将首尾节点衔接在一起
        head->left = pre;
        pre->right = head;
        return head;
    }
    vector<string> printTree(Node* root){
        string s;
        vector<string> levelOrder = levelOrderStr(root);
        return levelOrder;
    }
    void deleteTree(Node* root){
        if(root){
            deleteTree(root->left);
            deleteTree(root->right);
            root = nullptr;
        }
    }
    vector<string> levelOrderStr(Node* root){
        vector<Node*> cur;
        vector<Node*> next;
        if(!root)
            return vector<string> {};
        if(root)
            cur.push_back(root);
        vector<string> res;
//        vector<string> curEles;
        while(cur.size()){
            //标识next是否都为null
            int flag = 0;
            for(int i = 0;i < cur.size();i++){
                Node* node = cur[i];
                if(!node){
                    res.push_back("null");
                    continue;
                }
                else
                    res.push_back(to_string(node->val));
                //如果node的孩子不为null,则改变flag
                if(node->left || node->right)
                    flag++;
                next.push_back(node->left);
                next.push_back(node->right);
            }
            cur = next;
            if(flag == 0)
                cur.clear();
            next.clear();
        }
        return res;
    }
private:
    void inOrderHelper(Node* root,Node*& head,Node*& pre){
        if(root){
            inOrderHelper(root->left,head,pre);
            if(!pre){
                head = root;
                pre = root;
            }else{
                pre->right = root;
                root->left = pre;
                pre = root;
            }
            inOrderHelper(root->right,head,pre);
        }
    }
};

void printNode(Node* head){
    Node* cur = head;
    if(!cur){
        cout<<"[]"<<endl;
        return ;
    }
    cout<<"[";
    while(cur->right != head){
        cout<<cur->val<<"->";
        cur = cur->right;
    }
    cout<<cur->val<<"]"<<endl;
}


template<typename T>
void printVec(std::vector<T> l){
    std::string res = "[";
    std::string tmp = "";
    std::stringstream ss;
    for(auto ele : l){
        ss << ele;
        ss >> tmp;
        res += tmp + ",";
        ss.clear();
    }
    res = res.substr(0,res.size()-1);
    res += "]";
    std::cout<<res<<std::endl;
}

int main() {
    Node *root = new Node(4);
    Node *left = new Node(2);
    Node *right = new Node(5);
    Node* ll = new Node(1);
    Node* lr = new Node(3);
//    Node* rl = new Node(7);
//    Node* rr = new Node(9);
    root->left = left;
    root->right = right;
    left->left = ll;
    left->right = lr;
//    right->left = rl;
//    right->right = rr;
//    lr->left = new Node(3);
//    lr->right = new Node(5);
    vector<string> rootArr = Solution().levelOrderStr(root);
    printVec(rootArr);
    //[4,2,5,1,3]
    Node* head = Solution().treeToDoublyList(root);
    printNode(head);
    //result
    //[1->2->3->4->5]
}

File: offer_1_0036/java/offer_1_0036.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;

public class offer_1_0036 {
    class Node {
        public int val;
        public Node left;
        public Node right;

        public Node() {}

        public Node(int _val) {
            val = _val;
        }

        public Node(int _val,Node _left,Node _right) {
            val = _val;
            left = _left;
            right = _right;
        }
    };
    class Solution {
        public Node treeToDoublyList(Node root) {
            head = null;
            pre = null;
            if(root == null)
                return null;
            //中序遍历，保留前一个节点，修改当前节点的左子树指向(左子树已经遍历完可修改)和前一个节点的右子树指向
            inOrderHelper(root);
            //最后将首尾节点衔接在一起
            head.left = pre;
            pre.right = head;
            return head;
        }
        private Node head;
        private Node pre;
        private void inOrderHelper(Node root){
            if(root != null){
                inOrderHelper(root.left);
                if(pre == null){
                    pre = root;
                    head = root;
                }else{
                    pre.right = root;
                    root.left = pre;
                    pre = root;
                }
                inOrderHelper(root.right);
            }
        }
        public void printDoublyList(Node head){
            Node cur = head;
            if(cur == null){
                System.out.println("[]");
                return ;
            }
            System.out.printf("[");
            while(cur.right != head){
                System.out.printf("%d->",cur.val);
                cur = cur.right;
            }
            System.out.println(cur.val+"]");
        }
        public List<List<Integer>> levelOrder(Node root) {
            List<Node> cur = new ArrayList();
            List<Node> next = new ArrayList();
            if (root == null)
                return new LinkedList();
            if (root != null)
                cur.add(root);
            List<List<Integer>> res = new LinkedList();
            List<Integer> curEles = new LinkedList();
            while (cur.size() > 0) {
                for (int i = 0; i < cur.size(); i++) {
                    curEles.add(cur.get(i).val);
                    if (cur.get(i).left != null)
                        next.add(cur.get(i).left);
                    if (cur.get(i).right != null)
                        next.add(cur.get(i).right);
                }
                res.add(new LinkedList(curEles));
                cur = new LinkedList(next);
                curEles.clear();
                next.clear();
            }
            return res;
        }

        public List<String> levelOrderStr(Node root) {
            List<Node> cur = new ArrayList();
            List<Node> next = new ArrayList();
            if (root == null)
                return new LinkedList();
            if (root != null)
                cur.add(root);
            List<String> res = new LinkedList();
            while (cur.size() > 0) {
                int flag = 0;
                for (int i = 0; i < cur.size(); i++) {
                    if (cur.get(i) == null) {
                        res.add("null");
                        continue;
                    } else
                        res.add(String.valueOf(cur.get(i).val));
                    Node left = cur.get(i).left;
                    Node right = cur.get(i).right;
                    if (left != null || right != null)
                        flag++;
                    next.add(left);
                    next.add(right);
                }
                cur = new LinkedList(next);
                if (flag == 0)
                    cur.clear();
                next.clear();
            }
            return res;
        }

        public <T> void printListArr(List<List<T>> list) {
            for (int i = 0; i < list.size(); i++) {
                if (i == list.size() - 1) {
                    if (i == 0)
                        System.out.printf("[ ");
                    System.out.println(Arrays.toString(list.get(i).toArray(Integer[]::new)) + " ] ");
                } else {
                    if (i == 0)
                        System.out.printf("[ ");
                    System.out.printf(Arrays.toString(list.get(i).toArray(Integer[]::new)) + " , ");
                }
            }
        }
    }
    public static void main(String[] args) {
        offer_1_0036.Solution solution = new offer_1_0036().new Solution();
        offer_1_0036.Node root = new offer_1_0036().new Node(4);
        offer_1_0036.Node left = new offer_1_0036().new Node(2);
        offer_1_0036.Node right = new offer_1_0036().new Node(5);
        offer_1_0036.Node ll = new offer_1_0036().new Node(1);
        offer_1_0036.Node lr = new offer_1_0036().new Node(3);
//        offer_1_0036.Node rl = new offer_1_0036().new Node(7);
//        offer_1_0036.Node rr = new offer_1_0036().new Node(9);
        left.left = ll;
        left.right = lr;
//        right.left = rl;
//        right.right = rr;
        root.left = left;
        root.right = right;
//        lr.left = new offer_1_0036().new Node(3);
//        lr.left = new offer_1_0036().new Node(5);
        List<String> rootStr = solution.levelOrderStr(root);
        System.out.println(Arrays.toString(rootStr.toArray(String[]::new)));
        //[4, 2, 5, 1, 3, null, null]
        Node head = solution.treeToDoublyList(root);
        solution.printDoublyList(head);
        //result
        //[1->2->3->4->5]
    }
}
